[Wasm-GC] Support bulk array instructions

takikawa opened the pull request:
#### ea71f9290dfa4441fc1f5908e5bc41f189acfc35
<pre>
[Wasm-GC] Support bulk array instructions
<a href="https://bugs.webkit.org/show_bug.cgi?id=264057">https://bugs.webkit.org/show_bug.cgi?id=264057</a>

Reviewed by Justin Michaud.

Add support for bulk array operations from the Wasm GC proposal. These
allow bulk operations into GC arrays.

Adds spec tests from commit ff856bef3b717609e9b6854a1c78da80068322e9
of the GC proposal repo.

Also fixes a number of other related issues:
  * Fix incorrect element segment copying ranges for array.new_elem.
  * Fix handling of dropped segments in array.new_elem.
  * Minor fix to exception cases in B3 backend.
  * Fix registration of declared functions in the constant expression
    parser.
  * Adjust error messages
  * Fixed global import for i31refs.

* JSTests/wasm/gc-spec-tests/array_copy.wast.js: Added.
* JSTests/wasm/gc-spec-tests/array_fill.wast.js: Added.
* JSTests/wasm/gc-spec-tests/array_init_data.wast.js: Added.
* JSTests/wasm/gc-spec-tests/array_init_elem.wast.js: Added.
* JSTests/wasm/gc/array_new_elem.js:
(testAllElementSegmentKinds):
* JSTests/wasm/gc/arrays.js:
* JSTests/wasm/gc/bulk-array.js: Added.
(testArrayFill.doTest):
(testArrayFill):
(testArrayCopy.doTest):
(testArrayCopy):
(testArrayInitElem):
(testArrayInitData):
* JSTests/wasm/wasm.json:
* Source/JavaScriptCore/bytecode/BytecodeList.rb:
* Source/JavaScriptCore/llint/WebAssembly.asm:
* Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp:
(JSC::Wasm::B3IRGenerator::addArrayNewFixed):
(JSC::Wasm::B3IRGenerator::emitArrayNullCheck):
(JSC::Wasm::B3IRGenerator::addArraySet):
(JSC::Wasm::B3IRGenerator::addArrayFill):
(JSC::Wasm::B3IRGenerator::addArrayCopy):
(JSC::Wasm::B3IRGenerator::addArrayInitElem):
(JSC::Wasm::B3IRGenerator::addArrayInitData):
* Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
(JSC::Wasm::BBQJIT::addArrayFill):
(JSC::Wasm::BBQJIT::addArrayCopy):
(JSC::Wasm::BBQJIT::addArrayInitElem):
(JSC::Wasm::BBQJIT::addArrayInitData):
* Source/JavaScriptCore/wasm/WasmConstExprGenerator.cpp:
(JSC::Wasm::ConstExprGenerator::declaredFunctions const):
(JSC::Wasm::ConstExprGenerator::addRefFunc):
(JSC::Wasm::parseExtendedConstExpr):
* Source/JavaScriptCore/wasm/WasmConstExprGenerator.h:
* Source/JavaScriptCore/wasm/WasmExceptionType.h:
(JSC::Wasm::isTypeErrorExceptionType):
* Source/JavaScriptCore/wasm/WasmFunctionParser.h:
(JSC::Wasm::FunctionParser&lt;Context&gt;::parseExpression):
(JSC::Wasm::FunctionParser&lt;Context&gt;::parseUnreachableExpression):
* Source/JavaScriptCore/wasm/WasmIPIntGenerator.cpp:
(JSC::Wasm::IPIntGenerator::addArrayFill):
(JSC::Wasm::IPIntGenerator::addArrayCopy):
(JSC::Wasm::IPIntGenerator::addArrayInitElem):
(JSC::Wasm::IPIntGenerator::addArrayInitData):
* Source/JavaScriptCore/wasm/WasmInstance.cpp:
(JSC::Wasm::Instance::copyDataSegment):
(JSC::Wasm::Instance::copyElementSegment):
* Source/JavaScriptCore/wasm/WasmInstance.h:
* Source/JavaScriptCore/wasm/WasmLLIntBuiltin.h:
* Source/JavaScriptCore/wasm/WasmLLIntGenerator.cpp:
(JSC::Wasm::LLIntGenerator::addArrayFill):
(JSC::Wasm::LLIntGenerator::addArrayCopy):
(JSC::Wasm::LLIntGenerator::addArrayInitElem):
(JSC::Wasm::LLIntGenerator::addArrayInitData):
* Source/JavaScriptCore/wasm/WasmOperations.cpp:
(JSC::Wasm::JSC_DEFINE_JIT_OPERATION):
* Source/JavaScriptCore/wasm/WasmOperations.h:
* Source/JavaScriptCore/wasm/WasmOperationsInlines.h:
(JSC::Wasm::fillArray):
(JSC::Wasm::createArrayFromDataSegment):
(JSC::Wasm::createArrayFromElementSegment):
(JSC::Wasm::arrayNewElem):
(JSC::Wasm::arrayFill):
(JSC::Wasm::arrayCopy):
(JSC::Wasm::arrayInitElem):
(JSC::Wasm::arrayInitData):
* Source/JavaScriptCore/wasm/WasmSlowPaths.cpp:
(JSC::LLInt::WASM_SLOW_PATH_DECL):
* Source/JavaScriptCore/wasm/WasmSlowPaths.h:
* Source/JavaScriptCore/wasm/WasmTypeDefinition.cpp:
(JSC::Wasm::TypeInformation::signatureForLLIntBuiltin):
(JSC::Wasm::TypeInformation::TypeInformation):
* Source/JavaScriptCore/wasm/WasmTypeDefinition.h:
* Source/JavaScriptCore/wasm/js/JSWebAssemblyArray.cpp:
(JSC::JSWebAssemblyArray::fill):
(JSC::JSWebAssemblyArray::copy):
* Source/JavaScriptCore/wasm/js/JSWebAssemblyArray.h:
* Source/JavaScriptCore/wasm/js/WebAssemblyModuleRecord.cpp:
(JSC::WebAssemblyModuleRecord::initializeImports):
* Source/JavaScriptCore/wasm/wasm.json:

Canonical link: <a href="https://commits.webkit.org/272677@main">https://commits.webkit.org/272677@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ebb40f1a1cb019ccad063fb51561ddc4b3cb6540

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/32637 "3 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/11390 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/34491 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/35209 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/29502 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/13737 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/8577 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/29001 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/33076 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/9598 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/29157 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/8379 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/8501 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/29104 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/36545 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/27974 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/29659 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/29519 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/34634 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/32679 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/8625 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/6581 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/32498 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/10306 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/39126 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7591 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/9239 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/8259 "Passed tests") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/9227 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation reviewed (commented):
Can you add InPlace interpreter implementation too?

takikawa commented on Source/JavaScriptCore/wasm/WasmOperationsInlines.h:
```diff
@@ -312,6 +311,122 @@ inline void arraySet(Instance* instance, uint32_t typeIndex, EncodedJSValue arra
+
+// FIXME:
+// To be consistent it would make sense to include dst and src type indices, but they
+// are not necessary for operation and hits a limitation of BBQ JIT calls.
```
This comment is about working around what might be a bug in BBQJIT. I'm planning to file a proper bug report and put the bug link in here.

The basic issue seems to be that for a call to a WasmOperation, if the call arguments need to be allocated on the stack, it breaks in debug build mode on Linux. This wasn't an issue before because I don't think other operations exceed 6 arguments (x86-64 unix calling convention has 6 arg regs).

justinmichaud commented on Source/JavaScriptCore/wasm/js/JSWebAssemblyArray.cpp:
```diff
@@ -100,6 +100,78 @@ JSWebAssemblyArray::~JSWebAssemblyArray()
+        return;
+    // FIXME: support v128
+    default:
+        std::fill(m_payload64.begin() + offset, m_payload64.begin() + offset + size, value);
```
This is missing a write barrier, right? You should be able to get it to crash with collectContinuously=1

takikawa replied:
This should be ok actually, because reftypes are handled separately. Specifically on line 106 if the element type is a reftype then it will actually call the `set` method on each index instead of doing `std::fill`, which will use the write barrier path.

Same with the `copy` method below which also has a slow path for reftypes. 

There might be a write barrier issue with `array.init_elem` though, which I'll look into before merging.

takikawa replied:
For `array.init_elem`, I added a `vm.writeBarrier()` call to ensure the write barrier is used after the write into the array (though I wasn't able to trigger any failures with test cases for this). That should address everything that's left on this.

justinmichaud reviewed (approved):
r=me with the write barrier comment resolved. I think all three functions may be missing one, unless I am missing something

takikawa committed:
[Wasm-GC] Support bulk array instructions
https://bugs.webkit.org/show_bug.cgi?id=264057

Reviewed by Justin Michaud.

Add support for bulk array operations from the Wasm GC proposal. These
allow bulk operations into GC arrays.

Adds spec tests from commit ff856bef3b717609e9b6854a1c78da80068322e9
of the GC proposal repo.

Also fixes a number of other related issues:
  * Fix incorrect element segment copying ranges for array.new_elem.
  * Fix handling of dropped segments in array.new_elem.
  * Minor fix to exception cases in B3 backend.
  * Fix registration of declared functions in the constant expression
    parser.
  * Adjust error messages
  * Fixed global import for i31refs.

* JSTests/wasm/gc-spec-tests/array_copy.wast.js: Added.
* JSTests/wasm/gc-spec-tests/array_fill.wast.js: Added.
* JSTests/wasm/gc-spec-tests/array_init_data.wast.js: Added.
* JSTests/wasm/gc-spec-tests/array_init_elem.wast.js: Added.
* JSTests/wasm/gc/array_new_elem.js:
(testAllElementSegmentKinds):
* JSTests/wasm/gc/arrays.js:
* JSTests/wasm/gc/bulk-array.js: Added.
(testArrayFill.doTest):
(testArrayFill):
(testArrayCopy.doTest):
(testArrayCopy):
(testArrayInitElem):
(testArrayInitData):
* JSTests/wasm/wasm.json:
* Source/JavaScriptCore/bytecode/BytecodeList.rb:
* Source/JavaScriptCore/llint/WebAssembly.asm:
* Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp:
(JSC::Wasm::B3IRGenerator::addArrayNewFixed):
(JSC::Wasm::B3IRGenerator::emitArrayNullCheck):
(JSC::Wasm::B3IRGenerator::addArraySet):
(JSC::Wasm::B3IRGenerator::addArrayFill):
(JSC::Wasm::B3IRGenerator::addArrayCopy):
(JSC::Wasm::B3IRGenerator::addArrayInitElem):
(JSC::Wasm::B3IRGenerator::addArrayInitData):
* Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
(JSC::Wasm::BBQJIT::addArrayFill):
(JSC::Wasm::BBQJIT::addArrayCopy):
(JSC::Wasm::BBQJIT::addArrayInitElem):
(JSC::Wasm::BBQJIT::addArrayInitData):
* Source/JavaScriptCore/wasm/WasmConstExprGenerator.cpp:
(JSC::Wasm::ConstExprGenerator::declaredFunctions const):
(JSC::Wasm::ConstExprGenerator::addRefFunc):
(JSC::Wasm::parseExtendedConstExpr):
* Source/JavaScriptCore/wasm/WasmConstExprGenerator.h:
* Source/JavaScriptCore/wasm/WasmExceptionType.h:
(JSC::Wasm::isTypeErrorExceptionType):
* Source/JavaScriptCore/wasm/WasmFunctionParser.h:
(JSC::Wasm::FunctionParser<Context>::parseExpression):
(JSC::Wasm::FunctionParser<Context>::parseUnreachableExpression):
* Source/JavaScriptCore/wasm/WasmIPIntGenerator.cpp:
(JSC::Wasm::IPIntGenerator::addArrayFill):
(JSC::Wasm::IPIntGenerator::addArrayCopy):
(JSC::Wasm::IPIntGenerator::addArrayInitElem):
(JSC::Wasm::IPIntGenerator::addArrayInitData):
* Source/JavaScriptCore/wasm/WasmInstance.cpp:
(JSC::Wasm::Instance::copyDataSegment):
(JSC::Wasm::Instance::copyElementSegment):
* Source/JavaScriptCore/wasm/WasmInstance.h:
* Source/JavaScriptCore/wasm/WasmLLIntBuiltin.h:
* Source/JavaScriptCore/wasm/WasmLLIntGenerator.cpp:
(JSC::Wasm::LLIntGenerator::addArrayFill):
(JSC::Wasm::LLIntGenerator::addArrayCopy):
(JSC::Wasm::LLIntGenerator::addArrayInitElem):
(JSC::Wasm::LLIntGenerator::addArrayInitData):
* Source/JavaScriptCore/wasm/WasmOperations.cpp:
(JSC::Wasm::JSC_DEFINE_JIT_OPERATION):
* Source/JavaScriptCore/wasm/WasmOperations.h:
* Source/JavaScriptCore/wasm/WasmOperationsInlines.h:
(JSC::Wasm::fillArray):
(JSC::Wasm::createArrayFromDataSegment):
(JSC::Wasm::createArrayFromElementSegment):
(JSC::Wasm::arrayNewElem):
(JSC::Wasm::arrayFill):
(JSC::Wasm::arrayCopy):
(JSC::Wasm::arrayInitElem):
(JSC::Wasm::arrayInitData):
* Source/JavaScriptCore/wasm/WasmSlowPaths.cpp:
(JSC::LLInt::WASM_SLOW_PATH_DECL):
* Source/JavaScriptCore/wasm/WasmSlowPaths.h:
* Source/JavaScriptCore/wasm/WasmTypeDefinition.cpp:
(JSC::Wasm::TypeInformation::signatureForLLIntBuiltin):
(JSC::Wasm::TypeInformation::TypeInformation):
* Source/JavaScriptCore/wasm/WasmTypeDefinition.h:
* Source/JavaScriptCore/wasm/js/JSWebAssemblyArray.cpp:
(JSC::JSWebAssemblyArray::fill):
(JSC::JSWebAssemblyArray::copy):
* Source/JavaScriptCore/wasm/js/JSWebAssemblyArray.h:
* Source/JavaScriptCore/wasm/js/WebAssemblyModuleRecord.cpp:
(JSC::WebAssemblyModuleRecord::initializeImports):
* Source/JavaScriptCore/wasm/wasm.json:

Canonical link: https://commits.webkit.org/272677@main
